File: test/scenario.js
var sys    = require('sys'),
    path   = require('path'),
    http   = require('http'),
    assert = require('assert'),
    faye   = require('../build/faye-node');

Faye.Logging.logLevel = 'info';

AsyncScenario = Faye.Class({
  initialize: function(name) {
    this._name    = name;
    this._clients = {};
    this._inbox   = {};
    this._errors  = {};
    this._pool    = 0;
  },
  
  wait: function(time, Continue) {
    setTimeout(Continue, 1000 * time);
  },
  
  server: function(port, Continue) {
    this._endpoint = 'http://0.0.0.0:' + port + '/comet';
    this._server = new faye.NodeAdapter({mount: '/comet', timeout: 5});
    this._server.listen(port);
    Continue();
  },
  
  killServer: function(Continue) {
    if (this._server) this._server.close();
    this._server = undefined;
    setTimeout(Continue, 500);
  },
  
  httpClient: function(name, channels, Continue) {
    this._setupClient(new faye.Client(this._endpoint, {timeout: 8}), name, channels, Continue);
  },
  
  localClient: function(name, channels, Continue) {
    this._setupClient(this._server.getClient(), name, channels, Continue);
  },
  
  extendServer: function(stage, extension, Continue) {
    var object = {};
    object[stage] = extension;
    this._server.addExtension(object);
    Continue();
  },
  
  extendClient: function(name, stage, extension, Continue) {
    var object = {};
    object[stage] = extension;
    this._clients[name].addExtension(object);
    Continue();
  },
  
  listenForErrors: function(name, Continue) {
    var errors = this._errors[name] = [];
    this.extendClient(name, 'incoming', function(message, callback) {
      if (message.successful === false) errors.push(message.error);
      callback(message);
    }, Continue);
  },
  
  _setupClient: function(client, name, channels, Continue) {
    this._clients[name] = client;
    this._inbox[name]   = {};
    this._pool         += 1;
    
    Faye.each(channels, function(channel) {
      this.subscribe(name, channel);
    }, this);
    
    setTimeout(Continue, 500 * channels.length);
  },
  
  subscribe: function(name, channel, Continue) {
    var client = this._clients[name];
    
    this._lastSub = client.subscribe(channel, function(message) {
      var box = this._inbox[name];
      box[channel] = box[channel] || [];
      box[channel].push(message);
    }, this);
    
    setTimeout(Continue, 500);
  },
  
  cancelLastSubscription: function(Continue) {
    this._lastSub.cancel();
    setTimeout(Continue, 500);
  },
  
  publish: function(from, channel, message, Continue) {
    if (message instanceof Array)
      Faye.each(message, function(msg) {
        this._clients[from].publish(channel, msg);
      }, this);
    else
      this._clients[from].publish(channel, message);
    
    setTimeout(Continue, 500);
  },
  
  checkInbox: function(expectedInbox, Continue) {
    assert.deepEqual(this._inbox, expectedInbox);
    Continue();
  },
  
  checkErrors: function(name, expectedErrors, Continue) {
    assert.deepEqual(this._errors[name], expectedErrors);
    Continue();
  },
  
  finish: function(Continue) {
    Faye.each(this._clients, function(name, client) { client.disconnect() });
    var server = this._server;
    setTimeout(function() {
      server.close();
      Continue();
    }, 1000);
  },
});

SyncScenario = Faye.Class({
  initialize: function(name, block) {
    this._name = name;
    this._commands = [];
    this._scenario = new AsyncScenario();
    block.call(this);
  },
  
  run: function() {
    sys.puts('\n' + this._name);
    sys.puts('----------------------------------------------------------------');
    this._runNextCommand();
  },
  
  _runNextCommand: function() {
    if (this._commands.length === 0) return this._finish();
    
    var command = this._commands.shift(),
        method  = command[0],
        args    = Array.prototype.slice.call(command[1]),
        self = this;
    
    this._scenario[method].apply(this._scenario, args.concat(function() {
      self._runNextCommand();
    }));
  },
  
  _finish: function() {
    sys.puts('No errors; Shutting down server\n');
    this._scenario.finish(function() { SyncScenario.runNext() });
  }
});

['wait', 'server', 'killServer', 'httpClient', 'localClient',
 'subscribe', 'cancelLastSubscription', 'publish', 'checkInbox',
 'extendServer', 'extendClient', 'listenForErrors', 'checkErrors'].
forEach(function(method) {
  SyncScenario.prototype[method] = function() {
    this._commands.push([method, arguments]);
  }
});

Faye.extend(SyncScenario, {
  _queue: [],
  
  enqueue: function(name, block) {
    this._queue.push(new this(name, block));
  },
  
  runNext: function() {
    this.running = true;
    var scenario = this._queue.shift();
    if (!scenario) return;
    scenario.run();
  }
});

exports.run = function(name, block) {
  SyncScenario.enqueue(name, block);
  if (!SyncScenario.running) SyncScenario.runNext();
};

File: test/scenario.rb
$VERBOSE = false

module Scenario
  module ClassMethods
    def scenario(name, &block)
      define_method("test: #{name}", &block)
    end
  end
  
  def self.included(klass)
    klass.extend ClassMethods
  end
  
  def setup
    Thread.new { EM.run }
    while not EM.reactor_running?; end
    @scenario = AsyncScenario.new
    @commands = []
    @started = false
  end
  
  def teardown
    while EM.reactor_running?; end
  end
  
  def run(runner)
    @runner = runner
    super
  end
  
  def method_missing(sym, *args)
    @commands << [sym, args]
    EM.next_tick { run_next_command unless @started }
  end
  
  def run_next_command
    @started = true
    command = @commands.shift
    return finish if command.nil?
    begin
      @scenario.__send__(command.first, *command.last) do
        run_next_command
      end
    rescue Object => e
      @passed = false
      add_failure(e.message, e.backtrace)
      @runner.puke(self.class, self.name, e) if @runner.respond_to?(:puke)
      block.call
    end
  end
  
  def finish
    @scenario.finish { EM.stop }
  end
  
  class AsyncScenario
    include Test::Unit::Assertions
    
    def initialize
      @clients = {}
      @inbox   = {}
      @errors  = {}
      @pool    = 0
    end
    
    def wait(time, &block)
      EM.add_timer(time, &block)
    end
  
    def check_inbox(expected_inbox, &block)
      assert_equal expected_inbox, @inbox
      EM.next_tick(&block)
    end
    
    def check_errors(name, expected_errors, &block)
      assert_equal expected_errors, @errors[name]
      EM.next_tick(&block)
    end
    
    def server(port, &block)
      @endpoint = "http://0.0.0.0:#{port}/bayeux"
      @bayeux = Faye::RackAdapter.new(:mount => '/bayeux', :timeout => 30)
      Rack::Handler.get('thin').run(@bayeux, :Port => port) do |server|
        @server = server
        EM.next_tick(&block)
      end
    end
    
    def http_client(name, channels, &block)
      setup_client(Faye::Client.new(@endpoint), name, channels, &block)
    end
    
    def local_client(name, channels, &block)
      setup_client(@bayeux.get_client, name, channels, &block)
    end
    
    def extend_server(stage, extension, &block)
      object = Object.new
      def object.added
        @active = true
      end
      (class << object; self; end).send(:define_method, stage) do |*args|
        extension.call(*args) if @active
      end
      @bayeux.add_extension(object)
      block.call
    end
    
    def extend_client(name, stage, extension, &block)
      object = Object.new
      (class << object; self; end).send(:define_method, stage, &extension)
      @clients[name].add_extension(object)
      block.call
    end
    
    def listen_for_errors(name, &block)
      @errors[name] = []
      extend_client(name, :incoming, lambda { |message, callback|
        if message['successful'] == false
          @errors[name] << message['error']
        end
        callback.call(message)
      }, &block)
    end
    
    def setup_client(client, name, channels, &block)
      @clients[name] = client
      @inbox[name]   = {}
      @pool         += 1
      
      channels.each { |channel| subscribe(name, channel) }
      EM.add_timer(0.5 * channels.size, &block)
    end
    
    def subscribe(name, channel, &block)
      client = @clients[name]
      
      @last_sub = client.subscribe(channel) do |message|
        box = @inbox[name]
        box[channel] ||= []
        box[channel] << message
      end
      
      EM.add_timer(0.5, &block)
    end
    
    def cancel_last_subscription(&block)
      @last_sub.cancel
      EM.add_timer(0.5, &block)
    end
    
    def publish(from, channel, messages, &block)
      messages = [messages].flatten
      messages.each { |msg| @clients[from].publish(channel, msg) }
      EM.add_timer(2, &block)
    end
    
    def finish(&block)
      @clients.each { |id,c| c.disconnect }
      EM.add_timer(1) do
        @server.stop!
        block.call
      end
    end
  end
  
end

